Localization/extended_kalman_filter/extended_kalman_filter.py

```diff
@@ -10,44 +10,53 @@
 import math
 import matplotlib.pyplot as plt
 
-# Estimation parameter of EKF
-Q = np.diag([0.1, 0.1, np.deg2rad(1.0), 1.0])**2  # predict state covariance
-R = np.diag([1.0, 1.0])**2  # Observation x,y position covariance
+# Simulation parameters:
+
+# Process noise covariance
+Qsim = np.diag([
+    0.01, # variance of location on x-axis
+    0.01, # variance of location on y-axis
+    np.deg2rad(10.0), # variance of yaw-rate
+    0.01 # variance of velocity
+])**2
+
+# Observation noise covariance
+Rsim = np.diag([
+    0.5, # variance of measurement on x-axis
+    0.5 # variance of measurement on y-axis
+])**2
+
+
+# Estimation parameter of EKF:
+
+# Process noise covariance
+Q = Qsim * 1.5 # Qsim over-estimated by factor 1.5
+
+# Observation noise covariance
+R = Rsim * 1.5 # Rsim over-estimated by factor 1.5
 
-#  Simulation parameter
-Qsim = np.diag([1.0, np.deg2rad(30.0)])**2
-Rsim = np.diag([0.5, 0.5])**2
 
 DT = 0.1  # time tick [s]
 SIM_TIME = 50.0  # simulation time [s]
 
+
 show_animation = True
 
 
 def calc_input():
     v = 1.0  # [m/s]
     yawrate = 0.1  # [rad/s]
-    u = np.array([[v, yawrate]]).T
+    u = np.array([[v], [yawrate]])
     return u
 
 
 def observation(xTrue, xd, u):
 
-    xTrue = motion_model(xTrue, u)
-
-    # add noise to gps x-y
-    zx = xTrue[0, 0] + np.random.randn() * Rsim[0, 0]
-    zy = xTrue[1, 0] + np.random.randn() * Rsim[1, 1]
-    z = np.array([[zx, zy]]).T
-
-    # add noise to input
-    ud1 = u[0, 0] + np.random.randn() * Qsim[0, 0]
-    ud2 = u[1, 0] + np.random.randn() * Qsim[1, 1]
-    ud = np.array([[ud1, ud2]]).T
-
-    xd = motion_model(xd, ud)
+    xd = motion_model(xd, u) # assumes lack of noise
+    xTrue = motion_model(xTrue, u) + Qsim @ np.random.randn(4, 1)
```

Comment:
But the system within which the "robot" operates _is noisy_, no?
You would never find a vehicle be able to move on a "perfectly" circular path in a real environment. There will always be tiny perturbations.

The dead-reckoning however works on the theoretical and thus clean inputs alone, and hence diverges from the true state over time.

It's the "clean `x`" state, that's wrong, not the noisy one (at least in regards to the process noise, it should be reasonably clean from observation noise).

Reply:
In my assumption, xTrue is the true robot state without noise, which is a circle trajectory. however dead-reckoning sensor has noise, so xd is a trajectory with dead-reckoning with noise.

Reply:
With real-life effects like wheel slippage, carpet "springiness" and uneven floors a robot is unlikely to ever drive a perfect circle, as the one seen in these demos.

Note that the "effect of an unknown input" from [this](https://en.wikipedia.org/wiki/Kalman_filter#Example_application,_technical) …

> (note that there is no ![Bu](https://latex.codecogs.com/gif.latex?\inline&space;\dpi{120}&space;\small&space;{B}&space;u) term since we have no known control inputs. Instead, we assume that ![](https://latex.codecogs.com/gif.latex?\inline&space;\dpi{120}&space;\small&space;a_{k}) is the effect of an unknown input and ![](https://latex.codecogs.com/gif.latex?\inline&space;\dpi{120}&space;\small&space;{G}) applies that effect to the state vector) …

… also applies to scenarios with a known control input, referred to as process noise (i.e. `Qsim`).

So as long as `Qsim` is anything but a zero-filled matrix you will get a true state that's not following a perfect path (as you'd get by dead-reckoning its pure inputs).

---

Additionally the current version of the code feeds the EKF/UKF with (noisy) `x, y` coordinate measurements, while feeding the DR with (noise-added) `yawrate, velocity` inputs, which is somewhat akin to comparing apple and oranges. If you wanted a "perfect circle" for `xTrue` you would need to keep `Qsim` at (near-)zero and probably change the measurements for both to be `yawrate, velocity` by changing these:

```python
def observation_model(x):
    #  Observation Model
    H = np.array([
        [0, 0, 1, 0], // yaw rate
        [0, 0, 0, 1] // velocity
    ])
    z = H @ x
    return z

def observation(x, u):
    xTrue = motion_model(x, u) + Qsim @ np.random.randn(4, 1)
    z = observation_model(x) + Rsim @ np.random.randn(2, 1)

    uDR = u + Rsim @ np.random.randn(2, 1)
    xDR = motion_model(xDR, uDR)

    return xTrue, xDR, z
```

At which point you would get a perfect circle for `xTrue` and a noticeable drift for `xDR`.

The problem I would see with this however is that it would somewhat violate your goal `1.)`:

> 1. Easy to read for understanding each algorithm's basic idea.

… by obfuscating the relationships between `x`, `u`, `z`, `Q` and `R` in Kalman Filters. As you're effectively making use of the fact that you input `u` is of the same structure as your output `z`, hence allowing you to perform `u + Rsim @ np.random.randn(2, 1)`, which in the general case you cannot (due to likely differences in dimensionality and structure).

In contrast: the code in this PR expresses those relationships clearly and has them match the formulas found in text-books.

Reply:
>With real-life effects like wheel slippage, carpet "springiness" and uneven floors a robot is unlikely to ever drive a perfect circle, as the one seen in these demos.

I know that but I think it is not needed, because the purpose of this code is to understand the EKF algorithm, not the behavior of a robot in a real situation..